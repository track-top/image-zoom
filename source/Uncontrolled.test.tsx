import { act, fireEvent, render } from '@testing-library/react'
import React from 'react'
import { renderToString } from 'react-dom/server'
import Zoom from '../source'
import testA11y from '../testA11y'

test(`is accessible with defaults & <img />`, async () => {
  const { getByLabelText } = render(
    <main>
      <Zoom>
        <img alt="foo" src="foo.jpg" width="500" />
      </Zoom>
    </main>
  )
  const openTrigger = getByLabelText(`Zoom image`)
  expect(openTrigger).toBeVisible()

  expect(document.body).toMatchSnapshot()
  expect(await testA11y(document.body)).toEqual(true)

  fireEvent.click(openTrigger)

  // should do nothing
  fireEvent.click(openTrigger)

  const closeTrigger = getByLabelText(`Unzoom image`)
  expect(closeTrigger).toBeVisible()

  expect(document.body).toMatchSnapshot()
  expect(await testA11y(document.body)).toEqual(true)
})

test(`is accessible with custom open/close text & <img />`, async () => {
  const { container, getByLabelText } = render(
    <main>
      <Zoom closeText="Close me" openText="Open me">
        <img alt="foo" src="foo.jpg" width="500" />
      </Zoom>
    </main>
  )
  const openTrigger = getByLabelText(`Open me`)

  expect(openTrigger).toBeVisible()
  expect(await testA11y(container)).toEqual(true)

  fireEvent.click(openTrigger)

  const closeTrigger = getByLabelText(`Close me`)
  expect(closeTrigger).toBeVisible()

  expect(document.body).toMatchSnapshot()
  expect(await testA11y(document.body)).toEqual(true)
})

test(`zooms/unzooms with defaults & <img />`, () => {
  jest.useFakeTimers()

  const { getByLabelText, getByRole } = render(
    <Zoom>
      <img alt="foo" src="foo.jpg" width="500" />
    </Zoom>
  )
  const openTrigger = getByLabelText(`Zoom image`)
  expect(openTrigger).toBeVisible()

  expect(document.body).toMatchSnapshot()

  fireEvent.click(openTrigger)

  const closeTrigger = getByLabelText(`Unzoom image`)
  const modal = getByRole(`dialog`)

  expect(closeTrigger).toBeVisible()
  expect(document.body).toMatchSnapshot()

  fireEvent.click(closeTrigger)

  act(() => {
    jest.advanceTimersByTime(300)
  })

  expect(closeTrigger).not.toBeInTheDocument()
  expect(modal).not.toBeInTheDocument()
  expect(document.body).toMatchSnapshot()
})

test(`zooms/unzooms with custom open/close text & <img />`, async () => {
  jest.useFakeTimers()

  const { getByLabelText, getByRole } = render(
    <Zoom closeText="Close me" openText="Open me">
      <img alt="foo" src="foo.jpg" width="500" />
    </Zoom>
  )
  const openTrigger = getByLabelText(`Open me`)
  expect(openTrigger).toBeVisible()

  fireEvent.click(openTrigger)

  const closeTrigger = getByLabelText(`Close me`)
  const modal = getByRole(`dialog`)

  expect(closeTrigger).toBeVisible()
  expect(modal).toBeVisible()
  expect(document.body).toMatchSnapshot()

  fireEvent.click(closeTrigger)

  act(() => {
    jest.advanceTimersByTime(300)
  })

  expect(closeTrigger).not.toBeInTheDocument()
  expect(modal).not.toBeInTheDocument()
  expect(document.body).toMatchSnapshot()
})

test(`unzooms using ESC key`, () => {
  jest.useFakeTimers()

  const { getByLabelText, getByRole } = render(
    <Zoom>
      <img alt="foo" src="foo.jpg" width="500" />
    </Zoom>
  )
  const openTrigger = getByLabelText(`Zoom image`)
  expect(openTrigger).toBeVisible()

  fireEvent.click(openTrigger)
  const modal = getByRole(`dialog`)
  expect(modal).toBeVisible()

  expect(document.body).toMatchSnapshot()

  // should do nothing
  fireEvent.keyDown(document, { key: `ArrowLeft` })

  act(() => {
    jest.advanceTimersByTime(300)
  })

  expect(modal).toBeVisible()
  expect(document.body).toMatchSnapshot()

  fireEvent.keyDown(document, { key: `Escape` })

  act(() => {
    jest.advanceTimersByTime(300)
  })

  expect(modal).not.toBeInTheDocument()
  expect(document.body).toMatchSnapshot()

  fireEvent.click(openTrigger)
  expect(modal).toBeVisible()

  fireEvent.keyDown(document, { keyCode: 27 })

  act(() => {
    jest.advanceTimersByTime(300)
  })

  expect(modal).not.toBeInTheDocument()
  expect(document.body).toMatchSnapshot()
})

test(`unzooms on scroll`, () => {
  jest.useFakeTimers()

  const { getByLabelText, getByRole } = render(
    <Zoom>
      <img alt="foo" src="foo.jpg" width="500" />
    </Zoom>
  )
  const openTrigger = getByLabelText(`Zoom image`)
  expect(openTrigger).toBeVisible()

  fireEvent.click(openTrigger)
  const modal = getByRole(`dialog`)
  expect(modal).toBeVisible()

  expect(document.body).toMatchSnapshot()

  act(() => {
    window.dispatchEvent(new Event(`scroll`, {}))
  })

  act(() => {
    // run scroll again to emulate actual scroll events firing
    window.dispatchEvent(new Event(`scroll`, {}))
    jest.advanceTimersByTime(300)
  })

  expect(modal).not.toBeInTheDocument()
  expect(document.body).toMatchSnapshot()
})

test(`passes on original transform style`, () => {
  jest.useFakeTimers()

  const { getByLabelText, getByRole } = render(
    <Zoom wrapStyle={{ transform: `rotate(45deg)` }}>
      <img alt="foo" src="foo.jpg" width="500" />
    </Zoom>
  )
  const openTrigger = getByLabelText(`Zoom image`)
  expect(openTrigger).toBeVisible()

  expect(document.body).toMatchSnapshot()

  fireEvent.click(openTrigger)
  const modal = getByRole(`dialog`)
  expect(modal).toBeVisible()

  act(() => {
    jest.advanceTimersByTime(300)
  })

  const wrapEl = modal.querySelector<HTMLElement>(`[data-rmiz-modal-content]`)
  expect(wrapEl?.style?.transform).toContain(`rotate(45deg)`)

  expect(document.body).toMatchSnapshot()
})

test(`renders without browser environment`, () => {
  const html = renderToString(
    <Zoom>
      <img alt="foo" src="foo.jpg" width="500" />
    </Zoom>
  )

  document.body.innerHTML = html
  expect(document.body).toMatchSnapshot()
})

test(`render with different props`, () => {
  const html = renderToString(
    <Zoom
      overlayBgColorEnd="rgba(255, 255, 255, 0.95)"
      overlayBgColorStart="rgba(255, 255, 255, 0)"
      transitionDuration={400}
      wrapElement="span"
      zoomMargin={1}
      zoomZindex={2147483648}
    >
      <img alt="foo" src="foo.jpg" width="500" />
    </Zoom>
  )

  document.body.innerHTML = html
  expect(document.body).toMatchSnapshot()
})
