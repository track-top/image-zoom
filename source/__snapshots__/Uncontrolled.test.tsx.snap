// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`is accessible with custom open/close text & <img /> 1`] = `
<body>
  <div>
    <div
      data-rmiz-overlay="true"
    >
      <div
        tabindex="0"
      />
      <div
        aria-label="Zoomed image"
        aria-modal="true"
        role="dialog"
        style="background-color: rgba(255, 255, 255, 0.95); transition-duration: 300ms; z-index: 2147483647;"
        tabindex="-1"
      >
        <div
          data-rmiz-modal-content="true"
          style="height: 0px; left: 0px; top: 0px; transform: scale(Infinity) translate(0px, 0px); transition-duration: 300ms; width: 0px;"
        >
          <img
            alt="foo"
            src="foo.jpg"
            width="500"
          />
        </div>
        <button
          aria-label="Close me"
          data-rmiz-btn-close="true"
        />
      </div>
      <div
        tabindex="0"
      />
    </div>
  </div>
  <div />
  <div>
    <main>
      <div
        data-rmiz-wrap="hidden"
      >
        <img
          alt="foo"
          src="foo.jpg"
          width="500"
        />
        <button
          aria-label="Open me"
          data-rmiz-btn-open="true"
        />
      </div>
    </main>
  </div>
</body>
`;

exports[`is accessible with defaults & <img /> 1`] = `
<body>
  <div />
  <div />
  <div>
    <main>
      <div
        data-rmiz-wrap="visible"
      >
        <img
          alt="foo"
          src="foo.jpg"
          width="500"
        />
        <button
          aria-label="Zoom image"
          data-rmiz-btn-open="true"
        />
      </div>
    </main>
  </div>
</body>
`;

exports[`is accessible with defaults & <img /> 2`] = `
<body>
  <div>
    <div
      data-rmiz-overlay="true"
    >
      <div
        tabindex="0"
      />
      <div
        aria-label="Zoomed image"
        aria-modal="true"
        role="dialog"
        style="background-color: rgba(255, 255, 255, 0.95); transition-duration: 300ms; z-index: 2147483647;"
        tabindex="-1"
      >
        <div
          data-rmiz-modal-content="true"
          style="height: 0px; left: 0px; top: 0px; transform: scale(Infinity) translate(0px, 0px); transition-duration: 300ms; width: 0px;"
        >
          <img
            alt="foo"
            src="foo.jpg"
            width="500"
          />
        </div>
        <button
          aria-label="Unzoom image"
          data-rmiz-btn-close="true"
        />
      </div>
      <div
        tabindex="0"
      />
    </div>
  </div>
  <div />
  <div>
    <main>
      <div
        data-rmiz-wrap="hidden"
      >
        <img
          alt="foo"
          src="foo.jpg"
          width="500"
        />
        <button
          aria-label="Zoom image"
          data-rmiz-btn-open="true"
        />
      </div>
    </main>
  </div>
</body>
`;

exports[`passes on original transform style 1`] = `
<body>
  <div />
  <div />
  <div>
    <div
      data-rmiz-wrap="visible"
      style="transform: rotate(45deg);"
    >
      <img
        alt="foo"
        src="foo.jpg"
        width="500"
      />
      <button
        aria-label="Zoom image"
        data-rmiz-btn-open="true"
      />
    </div>
  </div>
</body>
`;

exports[`passes on original transform style 2`] = `
<body>
  <div>
    <div
      data-rmiz-overlay="true"
    >
      <div
        tabindex="0"
      />
      <div
        aria-label="Zoomed image"
        aria-modal="true"
        role="dialog"
        style="background-color: rgba(255, 255, 255, 0.95); transition-duration: 300ms; z-index: 2147483647;"
        tabindex="-1"
      >
        <div
          data-rmiz-modal-content="true"
          style="height: 0px; left: 0px; top: 0px; transform: scale(Infinity) translate(0px, 0px) rotate(45deg); transition-duration: 300ms; width: 0px;"
        >
          <img
            alt="foo"
            src="foo.jpg"
            width="500"
          />
        </div>
        <button
          aria-label="Unzoom image"
          data-rmiz-btn-close="true"
        />
      </div>
      <div
        tabindex="0"
      />
    </div>
  </div>
  <div />
  <div>
    <div
      data-rmiz-wrap="hidden"
      style="transform: rotate(45deg);"
    >
      <img
        alt="foo"
        src="foo.jpg"
        width="500"
      />
      <button
        aria-label="Zoom image"
        data-rmiz-btn-open="true"
      />
    </div>
  </div>
</body>
`;

exports[`render with different props 1`] = `
<body>
  <span
    data-rmiz-wrap="visible"
  >
    <img
      alt="foo"
      src="foo.jpg"
      width="500"
    />
    <button
      aria-label="Zoom image"
      data-rmiz-btn-open="true"
    />
  </span>
</body>
`;

exports[`renders without browser environment 1`] = `
<body>
  <div
    data-rmiz-wrap="visible"
  >
    <img
      alt="foo"
      src="foo.jpg"
      width="500"
    />
    <button
      aria-label="Zoom image"
      data-rmiz-btn-open="true"
    />
  </div>
</body>
`;

exports[`unzooms on scroll 1`] = `
<body>
  <div>
    <div
      data-rmiz-overlay="true"
    >
      <div
        tabindex="0"
      />
      <div
        aria-label="Zoomed image"
        aria-modal="true"
        role="dialog"
        style="background-color: rgba(255, 255, 255, 0.95); transition-duration: 300ms; z-index: 2147483647;"
        tabindex="-1"
      >
        <div
          data-rmiz-modal-content="true"
          style="height: 0px; left: 0px; top: 0px; transform: scale(Infinity) translate(0px, 0px); transition-duration: 300ms; width: 0px;"
        >
          <img
            alt="foo"
            src="foo.jpg"
            width="500"
          />
        </div>
        <button
          aria-label="Unzoom image"
          data-rmiz-btn-close="true"
        />
      </div>
      <div
        tabindex="0"
      />
    </div>
  </div>
  <div />
  <div>
    <div
      data-rmiz-wrap="hidden"
    >
      <img
        alt="foo"
        src="foo.jpg"
        width="500"
      />
      <button
        aria-label="Zoom image"
        data-rmiz-btn-open="true"
      />
    </div>
  </div>
</body>
`;

exports[`unzooms on scroll 2`] = `
<body>
  <div />
  <div />
  <div>
    <div
      data-rmiz-wrap="visible"
    >
      <img
        alt="foo"
        src="foo.jpg"
        width="500"
      />
      <button
        aria-label="Zoom image"
        data-rmiz-btn-open="true"
      />
    </div>
  </div>
</body>
`;

exports[`unzooms using ESC key 1`] = `
<body>
  <div>
    <div
      data-rmiz-overlay="true"
    >
      <div
        tabindex="0"
      />
      <div
        aria-label="Zoomed image"
        aria-modal="true"
        role="dialog"
        style="background-color: rgba(255, 255, 255, 0.95); transition-duration: 300ms; z-index: 2147483647;"
        tabindex="-1"
      >
        <div
          data-rmiz-modal-content="true"
          style="height: 0px; left: 0px; top: 0px; transform: scale(Infinity) translate(0px, 0px); transition-duration: 300ms; width: 0px;"
        >
          <img
            alt="foo"
            src="foo.jpg"
            width="500"
          />
        </div>
        <button
          aria-label="Unzoom image"
          data-rmiz-btn-close="true"
        />
      </div>
      <div
        tabindex="0"
      />
    </div>
  </div>
  <div />
  <div>
    <div
      data-rmiz-wrap="hidden"
    >
      <img
        alt="foo"
        src="foo.jpg"
        width="500"
      />
      <button
        aria-label="Zoom image"
        data-rmiz-btn-open="true"
      />
    </div>
  </div>
</body>
`;

exports[`unzooms using ESC key 2`] = `
<body>
  <div>
    <div
      data-rmiz-overlay="true"
    >
      <div
        tabindex="0"
      />
      <div
        aria-label="Zoomed image"
        aria-modal="true"
        role="dialog"
        style="background-color: rgba(255, 255, 255, 0.95); transition-duration: 300ms; z-index: 2147483647;"
        tabindex="-1"
      >
        <div
          data-rmiz-modal-content="true"
          style="height: 0px; left: 0px; top: 0px; transform: scale(Infinity) translate(0px, 0px); transition-duration: 300ms; width: 0px;"
        >
          <img
            alt="foo"
            src="foo.jpg"
            width="500"
          />
        </div>
        <button
          aria-label="Unzoom image"
          data-rmiz-btn-close="true"
        />
      </div>
      <div
        tabindex="0"
      />
    </div>
  </div>
  <div />
  <div>
    <div
      data-rmiz-wrap="hidden"
    >
      <img
        alt="foo"
        src="foo.jpg"
        width="500"
      />
      <button
        aria-label="Zoom image"
        data-rmiz-btn-open="true"
      />
    </div>
  </div>
</body>
`;

exports[`unzooms using ESC key 3`] = `
<body>
  <div />
  <div />
  <div>
    <div
      data-rmiz-wrap="visible"
    >
      <img
        alt="foo"
        src="foo.jpg"
        width="500"
      />
      <button
        aria-label="Zoom image"
        data-rmiz-btn-open="true"
      />
    </div>
  </div>
</body>
`;

exports[`zooms/unzooms with custom open/close text & <img /> 1`] = `
<body>
  <div>
    <div
      data-rmiz-overlay="true"
    >
      <div
        tabindex="0"
      />
      <div
        aria-label="Zoomed image"
        aria-modal="true"
        role="dialog"
        style="background-color: rgba(255, 255, 255, 0.95); transition-duration: 300ms; z-index: 2147483647;"
        tabindex="-1"
      >
        <div
          data-rmiz-modal-content="true"
          style="height: 0px; left: 0px; top: 0px; transform: scale(Infinity) translate(0px, 0px); transition-duration: 300ms; width: 0px;"
        >
          <img
            alt="foo"
            src="foo.jpg"
            width="500"
          />
        </div>
        <button
          aria-label="Close me"
          data-rmiz-btn-close="true"
        />
      </div>
      <div
        tabindex="0"
      />
    </div>
  </div>
  <div />
  <div>
    <div
      data-rmiz-wrap="hidden"
    >
      <img
        alt="foo"
        src="foo.jpg"
        width="500"
      />
      <button
        aria-label="Open me"
        data-rmiz-btn-open="true"
      />
    </div>
  </div>
</body>
`;

exports[`zooms/unzooms with custom open/close text & <img /> 2`] = `
<body>
  <div />
  <div />
  <div>
    <div
      data-rmiz-wrap="visible"
    >
      <img
        alt="foo"
        src="foo.jpg"
        width="500"
      />
      <button
        aria-label="Open me"
        data-rmiz-btn-open="true"
      />
    </div>
  </div>
</body>
`;

exports[`zooms/unzooms with defaults & <img /> 1`] = `
<body>
  <div />
  <div />
  <div>
    <div
      data-rmiz-wrap="visible"
    >
      <img
        alt="foo"
        src="foo.jpg"
        width="500"
      />
      <button
        aria-label="Zoom image"
        data-rmiz-btn-open="true"
      />
    </div>
  </div>
</body>
`;

exports[`zooms/unzooms with defaults & <img /> 2`] = `
<body>
  <div>
    <div
      data-rmiz-overlay="true"
    >
      <div
        tabindex="0"
      />
      <div
        aria-label="Zoomed image"
        aria-modal="true"
        role="dialog"
        style="background-color: rgba(255, 255, 255, 0.95); transition-duration: 300ms; z-index: 2147483647;"
        tabindex="-1"
      >
        <div
          data-rmiz-modal-content="true"
          style="height: 0px; left: 0px; top: 0px; transform: scale(Infinity) translate(0px, 0px); transition-duration: 300ms; width: 0px;"
        >
          <img
            alt="foo"
            src="foo.jpg"
            width="500"
          />
        </div>
        <button
          aria-label="Unzoom image"
          data-rmiz-btn-close="true"
        />
      </div>
      <div
        tabindex="0"
      />
    </div>
  </div>
  <div />
  <div>
    <div
      data-rmiz-wrap="hidden"
    >
      <img
        alt="foo"
        src="foo.jpg"
        width="500"
      />
      <button
        aria-label="Zoom image"
        data-rmiz-btn-open="true"
      />
    </div>
  </div>
</body>
`;

exports[`zooms/unzooms with defaults & <img /> 3`] = `
<body>
  <div />
  <div />
  <div>
    <div
      data-rmiz-wrap="visible"
    >
      <img
        alt="foo"
        src="foo.jpg"
        width="500"
      />
      <button
        aria-label="Zoom image"
        data-rmiz-btn-open="true"
      />
    </div>
  </div>
</body>
`;
